spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,499 @@
+    os.chdir("spam")
+
+# To visualize logs
+logger = logging.getLogger()
+logger.setLevel(logging.WARNING)
+
+# Show full columns for viewing data
+pd.set_option("display.max_colwidth", -1)
+
+# %% [markdown]
+# _Note:_ this tutorial differs from the labeling tutorial in that we use ground truth labels in the train split for demo purposes.
+# In practice, data slicing is agnostic to the _training labels_ used as inputs — you can use Snorkel-generated labels as inputs to this pipeline.
+
+# %%
+from utils import load_spam_dataset
+
+df_train, df_valid, df_test = load_spam_dataset(
+    load_train_labels=True, include_dev=False
+)
+
+df_train.head()
+
+
+# %% [markdown]
+# ## 1. Write slicing functions
+#
+# We leverage *slicing functions* (SFs) — an abstraction that shares syntax with *labeling functions*, which you should already be familiar with.
+# (If not, please see the [intro tutorial](https://github.com/snorkel-team/snorkel-tutorials/blob/master/spam/01_spam_tutorial.ipynb).)
+# A key difference: whereas labeling functions output labels, slicing functions output binary _masks_ indicating whether an example is in the slice or not.
+
+# %% [markdown]
+# In the following cells, we use the [`@slicing_function()`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/slicing/snorkel.slicing.slicing_function.html#snorkel.slicing.slicing_function) decorator to initialize an SF that identifies shortened links the spam dataset.
+# These links could redirect us to potentially dangerous websites, and we don't want our users to click them!
+# To select the subset of shortened links in our dataset, we write a regex that checks for the commonly-used `.ly` extension.
+#
+# You'll notice that the slicing function is noisily defined — it doesn't represent the ground truth for all short links.
+# Instead, SFs are often heuristics to quickly measure performance over important subsets of the data.
+
+# %%
+import re
+from snorkel.slicing import slicing_function
+
+
+@slicing_function()
+def short_link(x):
+    """Returns whether text matches common pattern for shortened ".ly" links."""
+    return bool(re.search(r"\w+\.ly", x.text))
+
+
+sfs = [short_link]
+
+# %% [markdown]
+# For our $n$ examples and $k$ slices in each split, we apply the SF to our data to create an $n \times k$ matrix. (So far, $k=1$).
+
+# %%
+from snorkel.slicing import PandasSFApplier
+
+applier = PandasSFApplier(sfs)
+S_train = applier.apply(df_train)
+S_valid = applier.apply(df_valid)
+S_test = applier.apply(df_test)
+
+# %% [markdown]
+# ### Visualize slices
+
+# %% [markdown]
+# With a utility function, [`slice_dataframe`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/slicing/snorkel.slicing.slice_dataframe.html#snorkel.slicing.slice_dataframe), we can visualize examples belonging to this slice in a `pandas.DataFrame`.
+
+# %%
+from snorkel.slicing import slice_dataframe
+
+short_link_df = slice_dataframe(df_valid, short_link)
```

Comment:
This feels a bit out of place in the narrative- you just applied this SF to all the data to create the `S_*` matrices, then they get totally dropped... it's a bit disorienting

Reply:
agreed, moved this down to when we actually use them